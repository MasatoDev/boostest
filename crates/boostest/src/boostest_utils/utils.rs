use anyhow::{anyhow, Result};
use std::path::{Path, PathBuf};

use glob::glob;
use std::collections::HashMap;
use std::ffi::OsStr;
use std::fs::{self, File};
use std::io::{self, prelude::*};

pub fn read(path: &Path) -> io::Result<String> {
    let mut f = File::open(path)?;
    let mut s = String::new();
    match f.read_to_string(&mut s) {
        Ok(_) => Ok(s),
        Err(e) => Err(e),
    }
}

pub fn read_matching_files(
    patterns: Vec<String>,
    out_file_name: &String,
) -> anyhow::Result<HashMap<PathBuf, String>> {
    let mut contents: HashMap<PathBuf, String> = HashMap::new();

    for pattern in &patterns {
        let result_glob = glob(pattern)?;

        for entry in result_glob {
            let path = entry?;

            if path
                .file_name()
                .unwrap_or(OsStr::new("none_file_name"))
                .to_string_lossy()
                .contains(out_file_name.as_str())
            {
                continue;
            }

            if path.is_file() {
                contents.insert(path.clone(), fs::read_to_string(path)?);
            }
        }
    }

    Ok(contents)
}

pub fn find_boostest_json_recursive(dir: PathBuf) -> anyhow::Result<PathBuf> {
    for entry in fs::read_dir(dir)? {
        let entry = entry?;
        let path = entry.path();

        if path.is_file() {
            if let Some(file_name) = path.file_name() {
                if file_name == "boostest.setting.json" {
                    return Ok(path);
                }
            }
        } else if path.is_dir() {
            if let Ok(found_path) = find_boostest_json_recursive(path) {
                return Ok(found_path);
            }
        }
    }

    Err(anyhow!("boostest.json not found"))
}

/**
 * Normalize and resolve the path.
 * parent: "hoge/a/b/c", relative: "../../d/e/f" => "hoge/a/d/e/f"
 */
pub fn normalize_and_resolve_path(parent: &Path, relative: &Path) -> Result<PathBuf, String> {
    let combined_path = parent.join(relative);
    let resolved_path = combined_path
        .components()
        .fold(PathBuf::new(), |mut acc, comp| {
            match comp {
                std::path::Component::ParentDir => {
                    acc.pop();
                }
                std::path::Component::CurDir => {}
                other => acc.push(other),
            }
            acc
        });

    if resolved_path.is_absolute() {
        Ok(resolved_path)
    } else {
        Err("The resolved path is not absolute.".to_string())
    }
}
