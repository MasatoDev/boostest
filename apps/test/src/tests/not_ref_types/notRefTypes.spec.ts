import { runSnapshotTest } from "../utils";

import { TsTypeLiteralLiteralObjectType } from "@/ts_types/ts_type_literal";
import { boostestDirectPromise } from "./boostest_output/boostestDirectPromise";
import { RefType, RefInterface } from "@/ts_types/utils";
import { Hoge } from "@/ts_types/literal";
import { boostestTuple } from "./boostest_output/boostestTuple";
import { boostestString } from "./boostest_output/boostestString";
import { boostestBoolean } from "./boostest_output/boostestBoolean";
import { boostestNumber } from "./boostest_output/boostestNumber";
import { boostestUnion } from "./boostest_output/boostestUnion";
import { boostestReference } from "./boostest_output/boostestReference";
import { boostestArray } from "./boostest_output/boostestArray";
import { boostestFunction } from "./boostest_output/boostestFunction";
import { boostestVoid } from "./boostest_output/boostestVoid";
import { boostestObject } from "./boostest_output/boostestObject";
import { boostestUnknown } from "./boostest_output/boostestUnknown";
import { boostestAny } from "./boostest_output/boostestAny";
import { boostestUndefined } from "./boostest_output/boostestUndefined";
import { boostestNull } from "./boostest_output/boostestNull";
import { boostestBigInt } from "./boostest_output/boostestBigInt";
import { boostestConditional } from "./boostest_output/boostestConditional";
import { boostestSymbol } from "./boostest_output/boostestSymbol";
import { boostestNamedTuple } from "./boostest_output/boostestNamedTuple";
import { boostestIntersection } from "./boostest_output/boostestIntersection";
import { boostestKeyof } from "./boostest_output/boostestKeyof";
import { boostestIndexAccessor } from "./boostest_output/boostestIndexAccessor";
import { boostestConstructor } from "./boostest_output/boostestConstructor";
import { boostestClass } from "./boostest_output/boostestClass";

describe("not ref types(direct) Tests", () => {
  runSnapshotTest(
    "Direct Promise",
    boostestDirectPromise<Promise<TsTypeLiteralLiteralObjectType>>(),
  );
  runSnapshotTest("Direct boolean", boostestBoolean<boolean>());
  runSnapshotTest("Direct string", boostestString<string>());
  runSnapshotTest("Direct number", boostestNumber<number>());
  runSnapshotTest("Direct bigint", boostestBigInt<bigint>());
  runSnapshotTest("Direct null", boostestNull<null>());
  runSnapshotTest("Direct undefined", boostestUndefined<undefined>());
  runSnapshotTest("Direct any", boostestAny<any>());
  runSnapshotTest("Direct unknown", boostestUnknown<unknown>());
  runSnapshotTest("Direct object", boostestObject<object>());
  runSnapshotTest("Direct void", boostestVoid<void>());
  runSnapshotTest(
    "Direct function",
    boostestFunction<(arg: RefType) => RefInterface>(),
  );
  runSnapshotTest("Direct array", boostestArray<string[]>());
  runSnapshotTest("Direct reference", boostestReference<RefType>());
  runSnapshotTest("Direct union", boostestUnion<string | number>());
  runSnapshotTest(
    "Direct conditional",
    boostestConditional<string extends number ? true : false>(),
  );
  runSnapshotTest("Direct symbol", boostestSymbol<symbol>());
  runSnapshotTest(
    "Direct tuple",
    boostestTuple<[string, number, any, RefType, RefInterface]>(),
  );
  runSnapshotTest(
    "Direct named tuple",
    boostestNamedTuple<
      [name: string, ver: number, ref: RefType, refInterface: RefInterface]
    >(),
  );
  runSnapshotTest(
    "Direct intersection",
    boostestIntersection<RefType & { name: string; age: number }>(),
  );
  runSnapshotTest("Direct keyof", boostestKeyof<keyof RefType>());
  runSnapshotTest(
    "Direct index accessor",
    boostestIndexAccessor<RefType["name"]>(),
  );
  runSnapshotTest(
    "Direct constructor",
    boostestConstructor<abstract new (...args: any) => Hoge>(),
  );
  runSnapshotTest("Direct class", boostestClass<typeof Hoge>());
});
