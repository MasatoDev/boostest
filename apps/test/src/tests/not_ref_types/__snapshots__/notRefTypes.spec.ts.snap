// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rename Tests Direct Promise matches snapshot 1`] = `"{"name":"test string data","age":10}"`;
